File: get_market_data.js
#!/usr/bin/env node
/**
 * 获取市场数据 - Node.js 包装器
 * 调用 Python 脚本获取 yfinance 数据
 */
const { execSync } = require('child_process');
const path = require('path');

// Python 路径（使用招财喵的 venv）
const PYTHON_PATH = 'C:\\Users\\Administrator\\Desktop\\zhaocai-miao\\.venv\\Scripts\\python.exe';
const SCRIPT_PATH = path.join(__dirname, 'get_market_data.py');

function getMarketData() {
  try {
    const output = execSync(`"${PYTHON_PATH}" "${SCRIPT_PATH}"`, {
      encoding: 'utf8',
      timeout: 60000
    });
    return JSON.parse(output);
  } catch (e) {
    console.error('获取市场数据失败:', e.message);
    return null;
  }
}

// 命令行调用
if (require.main === module) {
  const data = getMarketData();
  if (data) {
    console.log('市场数据获取成功:');
    console.log('BTC:', data.crypto.BTC);
    console.log('ETH:', data.crypto.ETH);
    console.log('黄金 (GLD):', data.gold.GLD, data.gold.unit);
    console.log('原油 (USO):', data.oil.USO, data.oil.unit);
    console.log('标普500:', data.indices['^GSPC'].price);
    console.log('纳斯达克:', data.indices['^IXIC'].price);
  }
}

module.exports = { getMarketData };

File: get_market_data.py
#!/usr/bin/env python3
"""
获取市场数据 - 供日知录文章使用
输出 JSON 格式，方便 Node.js 读取
"""
import yfinance as yf
import json
from datetime import datetime

def get_price(symbol, name=None):
    """获取单个标的的最新价格"""
    try:
        ticker = yf.Ticker(symbol)
        hist = ticker.history(period='1d')
        if hist.empty:
            return None
        price = hist['Close'].iloc[-1]
        return round(price, 2)
    except Exception as e:
        print(f"Error fetching {symbol}: {e}", flush=True)
        return None

def get_market_data():
    """获取所有市场数据"""
    data = {
        "update_time": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
        "crypto": {},
        "us_stocks": {},
        "gold": {},
        "oil": {},
        "indices": {}
    }
    
    # 加密货币
    crypto_pairs = [
        ("BTC-USD", "BTC"),
        ("ETH-USD", "ETH"),
    ]
    for symbol, name in crypto_pairs:
        price = get_price(symbol)
        if price:
            data["crypto"][name] = price
    
    # 美股
    us_stocks = [
        ("AAPL", "苹果"),
        ("TSLA", "特斯拉"),
        ("NVDA", "英伟达"),
        ("MSFT", "微软"),
        ("GOOGL", "谷歌"),
    ]
    for symbol, name in us_stocks:
        price = get_price(symbol)
        if price:
            data["us_stocks"][symbol] = {"name": name, "price": price}
    
    # 黄金期货 (GC=F)
    gc_price = get_price("GC=F")
    if gc_price:
        data["gold"] = {"GC=F": gc_price, "unit": "USD/盎司", "name": "黄金期货"}
    
    # 白银期货 (SI=F)
    si_price = get_price("SI=F")
    if si_price:
        data["silver"] = {"SI=F": si_price, "unit": "USD/盎司", "name": "白银期货"}
    
    # 原油：使用 USO ETF（暂无活跃期货合约）
    uso_price = get_price("USO")
    if uso_price:
        data["oil"] = {"USO": uso_price, "unit": "USD/桶", "note": "USO ETF，暂无活跃期货合约"}
    
    # 指数
    indices = [
        ("^GSPC", "标普500"),
        ("^DJI", "道琼斯"),
        ("^IXIC", "纳斯达克"),
        ("^OVX", "原油波动率指数"),
    ]
    for symbol, name in indices:
        price = get_price(symbol)
        if price:
            data["indices"][symbol] = {"name": name, "price": price}
    
    return data

if __name__ == "__main__":
    data = get_market_data()
    print(json.dumps(data, ensure_ascii=False, indent=2))
